python/tvm/relay/ir_pass.py

```diff
@@ -1,16 +1,292 @@
+        mod : tvm.relay.Module
+            The module that a certain optimization is performed on.
+
+        Returns
+        -------
+        mod : tvm.relay.Module
+            The updated module after applying this pass.
+        """
+        return _ir_pass.RunPass(self, mod)
+
+
+@register_relay_node
+class ModulePass(Pass):
+    """A pass that works on tvm.relay.Module.
+
+    Parameters
+    ----------
+    name : str
+        The module pass name.
+
+    opt_level : int
+        The optimization level of this module pass.
+
+    required : List[str]
+        The list of passes that are required by a module pass.
+
+    pass_func : Callable[(tvm.relay.Module, PassContext) -> tvm.relay.Module]
+        The callback function that sketches a certain optimization.
+    """
+
+    def __init__(self, name, opt_level, pass_func, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.CreateModulePass,
+                                            name, opt_level, required,
+                                            pass_func)
+
+
+@register_relay_node
+class FunctionPass(Pass):
+    """A pass that works on each tvm.relay.Function in a module.
+
+    Parameters
+    ----------
+    name : str
+        The function pass name.
+
+    opt_level : int
+        The optimization level of this function pass.
+
+    required : List[str]
+        The list of passes that are required by a function pass.
+
+    pass_func : Callable[(tvm.relay.Function, PassContext) ->
+                tvm.relay.Function]
+        The callback function that sketches a certain optimization.
+    """
+
+    def __init__(self, name, opt_level, pass_func, required=None):
+        self.__init_handle_by_constructor__(_ir_pass.CreateFunctionPass,
+                                            name, opt_level, required,
+                                            pass_func)
+
+
+@register_relay_node
+class SequentialPass(Pass):
+    """A pass that works on each tvm.relay.Function in a module.
+
+    Parameters
+    ----------
+    name : str
+        The sequential pass name.
+
+    opt_level : int
+        The optimization level of this sequential pass.
+
+    passes : List[Pass]
+        The pass candidates to be executed.
+
+    required : List[str]
+        The list of passes that are required by a sequential pass.
+
+    disabled : Optional[List[str]]
+        The list of passes that are disabled.
+    """
+
+    def __init__(self, name, opt_level, passes, required=None, disabled=None):
+        disabled = disabled if disabled else []
+        if not isinstance(disabled, (list, tuple)):
+            raise TypeError("disabled must be a list or tuple of pass names")
+        self.__init_handle_by_constructor__(_ir_pass.CreateSequentialPass,
+                                            name, opt_level, passes, required,
+                                            disabled)
+
+
+def create_module_pass(pass_name, opt_level, pass_func, required=None):
```

Comment:
We might want to consider enable a decorator pattern here.

```python
# Allow automatic derivation of pass name from the function name
@relay.ir_pass.module_pass(opt_level=2, required=[x, y, z])
def my_module_pass(mod, cx):
     ...

```

Reply:
changed